Add ClojureScript support (using cljc)

nberger opened the pull request:
Alternative version of #14, with bump to clojure 1.7.0 to use reader conditionals.

### Status

- [x] test.chuck.properties
- [x] test.chuck.generators (except double & regexes)
- [x] test.chuck.clojure-test
- [x] Implement a better test runner to avoid calling `(run-tests)` in each cljs test ns, and also to run tests in the browser.
- [x] Improve `lein test-all` (and/or add a new "test" alias) to compensate the issue of leiningen not picking up `.cljc` files. The namespaces have to be explicitly enumerated. See technomancy/leiningen#1940

nberger commented:
@gfredericks: this is WIP yet, but I wanted to open the PR so you (or anyone) can take a look early.

I tried different strategies to implement the `properties/for-all` macro for cljs, and the best alternative was to move the cljs version to a separate ns, so the users only need to use a reader conditional in the ns declaration. There's an [intermediate commit](https://github.com/nberger/test.chuck/commit/86d1e9212e9b7c684c95cdfda1043d5f69a8b999) using a separate macro name for cljs, but this way is clearly better IMO.

I'm working now on the `test.chuck.clojure-test` and its tests, and looks like it's a similar scenario with the three macros there.

nberger commented:
Also, I opened a separate PR with the idea to close the original, but perhaps I should just push it to #14. Please let me know which one do you think is better :)

gfredericks commented:
Re: `properties/for-all`: Shouldn't we be able to prevent users from having to use their own reader conditionals by using one ourselves in the actual impl?

nberger commented:
> Re: properties/for-all: Shouldn't we be able to prevent users from having to use their own reader conditionals by using one ourselves in the actual impl?

I thought we should, but couldn't find how. Found an explanation in the macros section of https://www.paren.com/posts/isomorphic-clojure-part-2-portable-code that makes much sense:

> Since ClojureScript macros are really Clojure macros, symbols are resolved in Clojure. For example, map would resolve to clojure.core/map instead of cljs.core/map. The only way around this faulty resolution is to quote and then unquote the symbol, like ~'map.

nberger commented:
And I understand Leon Grapenthin is referring to exactly that on his response (about [this post]( http://blog.fikesfarm.com/posts/2015-06-19-portable-macro-musing.html)) in [this thread](https://groups.google.com/d/msg/clojure/rOXCxIoiW0k/jiNw4rCDGzkJ)

gfredericks commented:
I don't mean that you'd necessarily have one `(defmacro for-all ...)`; you could, if necessary, have two of them in the same file, with a reader conditional at the top level. So same amount of code, just organized in a user-friendlier way. I might be wrong about how the reader conditional works for cljs macros though :/

I feel like there's gotta be a way to make it work one way or another, I just don't have time to think it through right this second.

nberger commented:
> I don't mean that you'd necessarily have one (defmacro for-all ...); you could, if necessary, have two of them in the same file, with a reader conditional at the top level. So same amount of code, just organized in a user-friendlier way. I might be wrong about how the reader conditional works for cljs macros though :/

Sadly, reader conditionals won't help in trying to have two `(defmacro for-all...)` in the same ns, one for cljs and one for clj: that's because `defmacro` is only valid inside of `#?(:clj ...)`.

gfredericks commented:
is there a runtime mechanism for a macro to detect if it's compiling clj code or cljs code?

nberger commented:
> is there a runtime mechanism for a macro to detect if it's compiling clj code or cljs code?

There is: `if (:ns &env)`. See [here](https://groups.google.com/forum/#!topic/clojurescript/iBY5HaQda4A). But from what I read, I understand it has some cons.

nberger commented:
I discarded that path when I read (from that thread):

> When you call a macro from a cljs file, and that macro calls another macro with a(:ns &env) test, it sees a Clojure &env, so it thinks is being compiled in a CLJ environment

I think we are in that situation: both `checking` and `for-all` need to know if they are being compiled in cljs, and both call `capture-reports` which also need to know that. As I am saying this, I think one simple solution to that issue is to inline the `capture-reports` macro. But I'm still not sure if that's it. We can try it :).

Anyways, I'd first finish with the current strategy, we are very close to have something usable with one gotcha: when used from cljc users will need a reader conditional to use the macros in cljs (which can be refactored (when used directly from a .cljs file, it's "just" that it's a different ns).

nberger commented:
`test.chuck.clojure-test` and `test.chuck.cljs-test` namespaces are working with tests (except `exception-detection-test` in cljs until we implement the cljs alternative for `(binding [clojure.test/*test-out* (java.io.StringWriter.)] ...)`

As of today, we have separate namespaces for cljs & clj macros, so for example in the case of `checking` and `for-all` the user will need a ns declaration like:

gfredericks commented:
This is sounding cool -- is it something that you're personally going to be using?

I haven't actually talked to anybody using test.check in cljs yet.

nberger commented:
Nice. I used test.check in a cljs project around 3 months ago, and by a quick look I see it could have made really good use of at least the `for-all` macro (had to do `and` on multiple truthy forms instead of multiple `is`) and `gen'/double` (used `gen/int`, so a few test cases were left out :smiley:).

I'm not going to use it right now, but will probably do in the next cljs project.

gfredericks commented:
okay cool; on a not-related-to-test.chuck note, I'd be interested if you would try out test.check version `0.8.0-RC2` to make sure it works and especially that it doesn't noticeably slow down your test suite.

nberger commented:
Sure, I'll do.

nberger commented:
I notice a slow down of around 10-20% in the cljs tests in that project, using `0.8.0-RC1` (couldn't find RC2) vs `0.7.0`. I can't say if it's a representative sample: it has only 1 test.check test. I measured with sizes 100 (1.3 vs 1.55 sec), 200 (5.2 vs 6.1 sec) and 1000 (25 vs 29 sec). The slow down seems consistent through the sample sizes.

nberger commented:
Of course, that's an ad-hoc benchmark. I can share more details with you, and/or try on different platforms, but better to do it somewhere else, to not lose the focus of this PR :)

gfredericks commented:
Thanks for the timing info. I'll be looking at the PR in more detail soon.

nberger commented:
Added the cljc version of exception-handling-test.

Merged in #16 to use it as the base implementation, those commits can be removed from here after #16 is merged into master.

@gfredericks: Do you have any idea in mind or preference about the cljsbuild profiles and test runner? Should we have a setup like that in test.check, with browser & node dev/adv profiles? Should we add it as part of this PR?

We currently have a very simple solution taken from core.match which expects every test ns to have the `(run-tests)` in the end and by using `:notify-command` the tests run via `lein cljsbuild auto adv`

gfredericks commented on project.clj:
```diff
@@ -3,12 +3,37 @@
+                    "com.gfredericks.test.chuck.clojure-test-test"
+                    "com.gfredericks.test.chuck.generators-test"
+                    "com.gfredericks.test.chuck.properties-test"
+                    "com.gfredericks.test.chuck.regexes-test"]
```
this `test` alias shouldn't be needed anymore, lein 2.5.2 supports cljc files.

nberger replied:
removed in 21dd3dd

gfredericks commented on project.clj:
```diff
@@ -3,12 +3,37 @@
             :url "http://www.eclipse.org/legal/epl-v10.html"}
-  :dependencies [[org.clojure/clojure "1.6.0"]
+  :dependencies [[org.clojure/clojure "1.7.0"]
+                 [org.clojure/clojurescript "0.0-3308"]
```
cljs has a new 1.7.something version series now

gfredericks replied:
although now that I think about it, cljs should probably be a dev dependency. Else anybody using test.chuck will now have cljs+etc on the classpath, which is complicated and sometimes bad.

nberger replied:
what about `:scope "provided"`? core.logic does it [that way](https://github.com/clojure/core.logic/blob/master/project.clj#L11). It means cljs jar will not be included in test.chuck uberjar. I'm not sure if it also means it won't be installed as a transitive dependency, but I think so. I'll check that now, and also bump to cljs 1.7.xx

nberger replied:
Bumped to clojurescript 1.7.48

Also added `:scope "provided"` to both clj & cljs. I checked and it works fine in that it doesn't install clojurescript as a transitive dependency when declaring the test.chuck dependency. And being strict, cljs is not a dev dependency, it's a "core" dependency, it's just that it's not always used, right?

But anyways, I'm okay if you still prefer to move it to `:dev`, just let me know.

gfredericks replied:
`:provided` sounds fine, especially if there's precedent for multi-platform libraries.

I don't care at all about test.check uberjars, I don't think that's a use case for anybody. I was only worried about transitive dependencies, both because it's annoying/noisy and also because there's more possibility for version conflicts e.g. if you're using any of the dependencies that cljs itself uses.

gfredericks commented on src/com/gfredericks/test/chuck/cljs_test.cljc:
```diff
@@ -0,0 +1,54 @@
+(ns com.gfredericks.test.chuck.cljs-test
```
I'm a bit iffy on the namespace name. test.check will be changing its namespace from `clojure.test.check.cljs-test` to `clojure.test.check.clojure-test` soon I think, so it might make sense for us to do the same thing.

nberger replied:
We have separate namespaces exactly because of the separation in test.check: as they have separate ns, we need to refer to `cljs-test.check.properties/for-all` and `clojure-test.check.properties/for-all` in our macro depending if we are compiling for clj or cljs, and we couldn't find the way to do that from a single macro.

Another solution is to have the two macros in the same namespace with different names. That's worst than the current scenario IMO, because it forces the user to use reader conditionals every time he was to use the macro, instead of only in the ns declaration.

So I guess we have two options:
1. Wait for the release of test.check where cljs-test ns is unified with clojure-test.
2. Keep the separate ns. When test.check changes its namespaces, test.chuck will follow (perhaps keeping the cljs-test ns with a deprecation notice until the next major release...)

gfredericks commented on src/com/gfredericks/test/chuck/clojure_test/impl.cljc:
```diff
@@ -0,0 +1,29 @@
+(defn ^:private not-falsey-or-exception?
+  [value]
+  (and value (not (instance? #?(:clj  Throwable
+                                :cljs js/Error)
```
oh this is interesting -- js let's you throw arbitrary objects, doesn't it? Does that mean that we can't currently distinguish a thrown not-`js/Error`-object from a returned object?

There are some changes in mind to test.check that would make this a bit cleaner, and I'm okay living with this for now, I just wanted to check if that's indeed the case.

gfredericks replied:
and test.check even on clj already has a converse problem where your test can't return (truthy) `Exception` objects without them being confused for thrown exceptions, so this is definitely okay for now.

nberger replied:
Thanks for pointing this out. I think you are correct. Related to that, we are not being able to report a `:fail`ed test due to an exception in an assertion as a distinct result from an `:error`ed test due to an exception not in an assertion. That can be seen in [our test](https://github.com/gfredericks/test.chuck/pull/15/files#diff-2eaa93a3b0a438ecb9060dfbfb3a393cR33). It seems to have its roots in [test.check not distinguishing between failure and error](https://github.com/clojure/test.check/blob/master/src/main/clojure/clojure/test/check/clojure_test.clj#L147-L153), which is not a trivial topic given shrinking is involved.

gfredericks commented on src/com/gfredericks/test/chuck/generators.cljc:
```diff
@@ -78,7 +82,8 @@
                     gens (map second pairs)]
-                `(for [[~@names] (gen/tuple ~@gens)
+                `(for [[~@names]
+                       (~'gen/tuple ~@gens)
```
I'm still a bit confused about how cljs macros fit in everywhere, but my guess is that this is dangerous -- does it not assume that every namespace that calls this macro has done `(:require [clojure.test.check.generators :as gen])`?

I can see how it would be harder to do generally though, because the generators namespace is different on the different platforms, is that right? If that's the biggest blocker, I've plans to rename the cljs namespaces in test.check, so I could expidite that.

nberger replied:
Yes, you are correct. This is wrong, has to be fixed in a different way. We can't use the ns alias with quote-unquote, at least it should be the full ns name.

My worry is that we are going to play the "separate macros for clj and cljs" dance here too.

gfredericks replied:
but is it true that if test.check had unified namespace names the problem would go away?

nberger replied:
yes :)

gfredericks commented on src/com/gfredericks/test/chuck/generators.cljc:
```diff
@@ -144,7 +149,8 @@
   [elements]
-  (for [bools (apply gen/tuple (repeat (count elements) gen/boolean))]
+  (for [bools (apply gen/tuple
+                     (repeat (count elements) gen/boolean))]
```
I won't be a stickler for this, but I'd prefer to avoid whitespace changes; don't mind a separate PR for just whitespace though.

nberger replied:
Whitespace change removed. This is especially important given this is already a quite big PR.

gfredericks commented on src/com/gfredericks/test/chuck/properties/macros.clj:
```diff
@@ -0,0 +1,18 @@
+                    (gen'/for ~bindings
+                      (with-meta
+                        ~(zipmap quoted-names bound-names)
+                        {::for-all-bindings-map true}))]
```
this metadata is probably unnecessary; I can't see any other reference to that keyword. But that could be a separate commit.

nberger replied:
Yes, it's not being used anywhere in the project. When it was [introduced](https://github.com/gfredericks/test.chuck/commit/066d7bffc1a38c8842e5560d28c822dedea88b9c) it was supposed to be used by tooling. We could try asking the guy who did that commit :P, or if he doesn't remember if he ended using it, we could just remove it and see if someone complains. I couldn't find a reference to it [anywhere in github](https://github.com/search?q=for-all-bindings-map&type=Code&utf8=%E2%9C%93)

But yes, I think it's better to remove it in a separate commit, because we would remove it from clj (already released) and cljs side, right?

gfredericks replied:
it might have been my idea; I had forgotten about the tooling angle. Probably just keep it then.

gfredericks commented on test/com/gfredericks/test/chuck/clojure_test_test.cljc:
```diff
@@ -0,0 +1,47 @@
+
+#?(:cljs (set! *print-fn* js-print))
+
+#?(:cljs (run-tests))
```
doesn't test.check have a global test-runner file that handles stuff like this and runs the other files?

nberger replied:
Yes. I started with those explicit `(run-tests)` because it was very easy to start rolling, but it's definitely something to improve. I'll take the test.check approach.

nberger replied:
Instead of the test.check approach, what do you think about using https://github.com/bensu/doo? I gave it a try and it's working in my "doo" branch: https://github.com/nberger/test.chuck/tree/doo

The advantage I see is basically that there's less boilerplate, and we get more runners (slimer.js and rhino) for "free".

gfredericks commented on test/com/gfredericks/test/chuck/properties_test.cljc:
```diff
@@ -0,0 +1,49 @@
+                     [cljs.test.check :as t.c]
+                     [cljs.test.check.generators :as gen]
+                     [cljs.test.check.properties :include-macros true]
+                     [cljs.test.check.cljs-test :refer-macros [defspec]])))
```
this makes me want to just go release the renamed test.check namespaces right now.

nberger replied:
Yes, please do it :). Btw is there a test.check branch or patch somewhere doing that?

gfredericks replied:
working on it; I'll most likely make another RC release shortly

gfredericks replied:
Okay there's a branch up called `cljs-namespace-renaming` that you ought to be able to play with. I'm going to try it on one of my own projects and release an RC if it works

nberger replied:
Cool, I'll give it a try. I see it's already in master, so I guess it worked on your project.

gfredericks replied:
`0.8.0-RC3` is in the oven and should be available within 20 minutes or so. I fixed the problem with latest cljs, so you shoud be able to use `1.7.48` as well.

nberger replied:
Thanks. I'm making progress in using the unified namespaces: generators-test is working, properties-test is closer to be working. I should be able to push something or report issues soon.

gfredericks commented:
okay test-check `0.8.0-RC2` is out. You might have to bump cljs down to 1.7.28 to get it to work though. Something about 1.7.48 is bad with the new test.check and I haven't figured out what yet.

nberger commented:
@gfredericks: tests are working with `0.8.0-RC3`, using the unified namespaces.

__But__: I still had to keep the two separate namespaces `test.chuck.cljs-test` and `test.chuck.clojure-test`. That's because of at least two things:

1. the `capture-reports` macros which still have some differences between clj & cljs
2. the call to cljs.test/report & clojure.test/report at the end of the `checking` macro.

I'll if I can find a way to overcome this. But for now, it forces us to have separate `checking` and `for-all` macros, thus separate namespaces.

nberger commented:
Also: I merged the branch for switching to bensu/doo for the test runner. It includes a small addition to the README explaining how to run the tests. Let me know whether you like it or if you still prefer to take the runners from test.check

gfredericks commented:
what is it about the `test.chuck.{clojure,cljs}-test` namespaces that prevents handling the differences with reader conditionals?

gfredericks commented:
I'm not opposed to `bensu/doo` if it's a solid library and does reduce boilerplate

nberger commented:
> what is it about the test.chuck.{clojure,cljs}-test namespaces that prevents handling the differences with reader conditionals?

It's the need to refer to different symbols depending on whether it's compiling for cljs or clj. This cannot be done using reader conditionals inside of a macro, because it will always go through the `:clj` branch.

There's one thing I haven't tried: The `(:ns &env)` runtime trick (see https://github.com/gfredericks/test.chuck/pull/15#issuecomment-127768642). I'll give it a whirl and let you know.

About lein-doo: I think it's solid, and being just dev tooling we can change it later if we want to, so let's use it :)

nberger commented:
@gfredericks: Made the namespace unification by using `(:ns &env)`. It's working. I had to inline the `capture-reports` macro, and there's a bit of repeated code inside of the macros, but hey, it works! :)

I mean, we are free to refactor the macro code later if we want to, but at least we got rid of the separate namespaces. Thanks for insisting on that.

Let me know how it looks to you.

nberger commented:
@gfredericks: moved test.check to the global dependencies vector. We want it pulled as a transitive dependency, so a project would just declare `test.chuck` to get both, right?

Btw, this change could also be done in master. Do you want me to create a separate PR to do that on the current master?

Also, do you want me to try to extract smaller PR's from this bigger one? Probably won't be able to extract many, but we already have #16 for example.

gfredericks commented:
> It's the need to refer to different symbols depending on whether it's compiling for cljs or clj. This cannot be done using reader conditionals inside of a macro, because it will always go through the `:clj` branch.

Dangit I always forget about that

gfredericks commented:
I don't think projects would realistically not declare their own test.check dependency, but I suppose they could. It's probably good to have test.check as a real dep anyhow so that we have the version set and users can see conflicts.

gfredericks commented:
so with the `(:ns &env)` approach there was a downside with writing macros that expand to the macro that it appears in, is that right? Is there a workaround where the macro-author could somehow set `:ns` explicitly?

nberger commented:
> so with the (:ns &env) approach there was a downside with writing macros that expand to the macro that it appears in, is that right? Is there a workaround where the macro-author could somehow set :ns explicitly?

Yes, seems like that's the downside. Don't know, will investigate more about the :ns thing.

nberger commented:
See this excerpt from https://groups.google.com/d/msg/clojurescript/iBY5HaQda4A/SuVYqNRH8gwJ:

> macros that want to emit cross-platform code should emit forms like `(if-cljs cljs-form clj-form) -- notice the backtick -- and even if they are called from other macros, the ultimate expansion takes place in the correct target context (I hope).  

We are emiting forms like that (with backtick), so we should be fine to call this macros from other macros without the need to set :ns.

I don't see someone building macro sugar on top of the test.chuck macro sugar, but it can be done (test.chick project coming...). But what's importante here is that test.check could use a similar strategy for the underlying macros, and test.chuck should be able to call them. Should we start the "refactor to cljc" branch in test.check? :)

gfredericks commented:
I do have plans to move test.check to cljc but I thought I might wait till at least the next release to delay requiring a 1.7 upgrade from users.

nberger commented:
Sounds reasonable. Would that be the 0.8.1 release?

gfredericks commented:
Could we sidestep the cross-platfrom macro problem by factoring most of the functionality out of the macro and into functions?

nberger commented:
> Could we sidestep the cross-platfrom macro problem by factoring most of the functionality out of the macro and into functions?

I'm not sure if we can refactor into functions in this case, because we do the wrap of the body into a call to a (platform-specific) `binding`... I tried to extract a function or a macro just to generate the binding vector, but failed. If you can do that, please enlighten me :)

Anyways, your call to refactor made me remember that I inlined the `capture-reports` macro in the middle of incorporating the `if-cljs` macro... So I stepped back and extracted that macro again, and looks much better now. It's in the two commits I just pushed.

Next (easy) step is to extract a macro for the [`report-when-failing` call](https://github.com/gfredericks/test.chuck/pull/15/files#diff-f50935dcebcdd3fc4abf39442113df09R53) which is the same for both platforms now.

Let me know if you see something I am missing or a better way to do this.

nberger commented:
Extracted the `qc-and-report-when-failing` macro. The platform-specific stuff is now confined to `capture-reports` and (a much simpler) `checking` macro.

nberger commented:
@gfredericks:

I just merged in master, and implemented `gen/datetime` for cljs using cljs-time.

As we are still doing reviews I'm keeping the raw commits. We can rebase to master (and squash commits) anytime. I'd do it once we are ready to merge, but that's your call.

gfredericks commented:
so is this all done to the best of your knowledge? I'm curious if the latest test.check release has been sufficient for this task; if so I'll probably consider that justification for a full release.

nberger commented:
I think so. With regard to namespaces and macros, the latest test.check
release has been a great help.

gfredericks commented:
planning on looking at this in detail in the next few days. thanks again for all the work.

nberger commented:
Ok, thanks.

I just tried to sync with master but I found that a test started to fail. I'll wait for the resolution of #23 before doing the sync.

gfredericks commented:
If I don't get to this sooner I will definitely definitely get to it on Tuesday.

nberger commented:
Ok, thanks.

I merged this with master.

I also pushed a [cljc-rebased](https://github.com/nberger/test.chuck/tree/cljc-rebased) branch to my repo with all of this rebased to master, with the commits squashed to just a few. I'm keeping the two branches so as to not interfere with the review, but I think we should merge something like cljc-rebased instead of this :)

gfredericks commented on README.md:
```diff
@@ -234,6 +234,23 @@ More details in [this blog post](http://blog.colinwilliams.name/blog/2015/01/26/
+
+For cljs [bensu/doo](https://github.com/bensu/doo) is configured so the tests can be run with:
+
+    $ lein doo node node-test
```
when I do this it seems to do everything except actually run the tests; if I tweak a test to be failing, I don't see any indication of that.

nberger replied:
I have the exact same issue when using node v0.10.28. It works fine with v0.12.7. Could you please try on node 0.12.x?

nberger replied:
Sent a [PR](https://github.com/bensu/doo/pull/27) to bensu/doo to clarify the minimum required node version in their README. Do you think we should add a note here, too?

gfredericks replied:
yeah, if it's definitely a requirement then a comment would be good; I haven't had a chance to try it out yet

nberger replied:
Comment was added

gfredericks replied:
apparently node is on v4.0 now??? anyhow I ran the tests with that and it worked so cool

gfredericks commented on src/com/gfredericks/test/chuck/clojure_test.cljc:
```diff
@@ -0,0 +1,81 @@
+         ~@body)
+       (binding [clojure.test/report #(swap! reports# conj %)]
+         ~@body))
+     @reports#))
```
Can we avoid `if-cljs` with something like this?

``` clojure
(defn capture-reports*
  [reports-atom func]
  #? (:clj
      (binding [clojure.test/report #(swap! reports-atom conj %)]
        (func))
      :cljs
      (binding [*chuck-captured-reports* reports-atom
                cljs.test/*current-env* (cljs.test/empty-env ::chuck-capture)]
        (func))))

(defmacro capture-reports
  [& body]
  `(let [reports# (atom [])]
     (capture-reports* reports# (fn [] ~@body))
     @reports#))
```

nberger replied:
Yessss! of course we can, thanks :)

Done in bce3f46c67d0edb8ed281561e2a96fd9d44dae0f

gfredericks commented on src/com/gfredericks/test/chuck/clojure_test.cljc:
```diff
@@ -0,0 +1,81 @@
+     (let [final-reports# (atom [])]
+       (qc-and-report-exception final-reports# ~tests ~bindings ~@body)
+       (doseq [r# @final-reports#]
+         (clojure.test/report r#))))))
```
I guess if we aim to get rid of `if-cljs` we have to figure this one out too (I hadn't noticed it before).

Looking at the impl of both of the `testing` macros, I think the following should work:

``` clojure
(defn -testing
  [name func]
  (#?(:clj clojure.test/testing :cljs cljs.test/testing) name (func)))

(defmacro checking
  "..."
  [name tests bindings & body]
  (-testing ~name
    (fn []
      (let [final-reports# (atom [])]
        (qc-and-report-exception final-reports# ~tests ~bindings ~@body)
        (doseq [r# @final-reports#]
          (clojure.test/report r#))))))
```

gfredericks replied:
the key detail is that the name arg to the `testing` macros does not have to be a string literal

nberger replied:
Thanks, thanks, thanks! It worked!

I only needed to add syntax-quote, and introduce a similar `-report` fn. Done in fd1f40571b1fc04154c6f646851dc323d04807c5

gfredericks commented:
I hate to ask the same question every three weeks, but I couldn't figure this out from the comment history:

Why do we have a separate `.clojure-test.impl` namespace?

gfredericks commented on src/com/gfredericks/test/chuck/generators.cljc:
```diff
@@ -1,14 +1,16 @@
 
 ;; Hoping this will be in test.check proper:
 ;; http://dev.clojure.org/jira/browse/TCHECK-15
+#?(:clj
```
this isn't necessary is it? I didn't see that around macros in the other files

nberger replied:
Yep, it's not needed. Thanks, I just removed it.

gfredericks commented on src/com/gfredericks/test/chuck/generators.cljc:
```diff
@@ -80,7 +82,7 @@
                     names (map first pairs)
                     gens (map second pairs)]
-                `(for [[~@names] (gen/tuple ~@gens)
+                `(for [[~@names] (~'clojure.test.check.generators/tuple ~@gens)
```
why is this necessary?

nberger replied:
Hummm, not sure why we left it that way... Perhaps we needed something like that when clojure.test.check had separate namespaces for clj & cljs, but it's not the case anymore. Just changed it to use the `gen` alias

gfredericks commented:
Just did a full scan of the changes, looking pretty good aside from the comments I made, so I think we're close. Will also need another merge from master.

nberger commented:
Thanks for the thorough review. I just merged with master, and will take a look into the comments ASAP.

nberger commented:
> Why do we have a separate .clojure-test.impl namespace?

Good catch! The separation was needed when we had separate macros and namespaces for clj & cljs, so as to reuse that code from both ns, but that's not the case anymore. I just moved the code to `.clojure-test` and removed `.clojure-test.impl`

nberger commented:
I think I've answered everything :)

I've also rebased my [cljs-rebased](https://github.com/nberger/test.chuck/tree/cljc-rebased) branch again, which is based on master and includes everything from here. It's only 4 commits versus 48 here, so we should probably make this branch to point to cljc-rebased before merging :)

To be sure both branches have the same content, I'm running `git diff cljc-rebased cljc`. github [compare view](https://github.com/nberger/test.chuck/compare/cljc...cljc-rebased) doesn't work in this case.

gfredericks commented on src/com/gfredericks/test/chuck/properties.cljc:
```diff
@@ -81,8 +80,8 @@
-    `(prop/for-all [{:syms [~@bound-names]}
-                    (gen'/for ~bindings
+    `(clojure.test.check.properties/for-all [{:syms [~@bound-names]}
+                    (com.gfredericks.test.chuck.generators/for ~bindings
```
this didn't need to change either?

nberger replied:
Yes, I changed this and a couple more places where an alias could be used in 5ad5d11

gfredericks commented on test/com/gfredericks/test/chuck/exception_handling_test.cljc:
```diff
@@ -0,0 +1,63 @@
+           :cljs
+           (binding [*current-env* (test/empty-env)]
+             (capture-test-var #'this-test-should-crash-and-be-caught)
+             (:report-counters *current-env*)))]
```
the logic represented by this `#?(...)` in duplicated in the next test, and is hard to read in this context. Could it be extracted into a function whose name describes what it's doing? (which I assume is running the test and returning the results)

nberger replied:
Done in a01278f

gfredericks commented on test/com/gfredericks/test/chuck/properties_test.cljc:
```diff
@@ -0,0 +1,34 @@
+               (tc/quick-check 100
+                 (prop'/for-all [x gen/int]
+                   #?(:clj  (/ 4 0)
+                      :cljs (js/Error. "Oops"))))))))
```
you're not throwing the error here

gfredericks replied:
there's gotta be a platform-agnostic way to make something crash :/ what does `(subs nil 1 2)` do in cljs?

nberger replied:
Yes, there must be some way, for sure. I'll take a look tomorrow

nberger replied:
Changed it to (in a9c6904):

```clojure
(checking "subs on any string or nil" 100
  [s (gen/one-of [gen/string (gen/return nil)])]
  ;; going for uncaught-error-not-in-assertion here
  (let [res (subs s 1 2)]
    (is res)))
```

But also did something similar in exception-handling-test and now I get `{:pass 1, :fail 1, :error 0}` instead of `{:pass 0, :fail 1, :error 0}` in the `exception-detection-test` in cljs. In clj works fine as before, returning `{:pass 0, :fail 1, :error 0}`.

Not sure why this is happening, but I don't want to fix it just by changing it to a different expression or something like that... I want to understand it first.

Any ideas?

nberger replied:
Cljs tests are running in circle now, failing of course because of this issue :/

https://circleci.com/gh/nberger/test.chuck/26

gfredericks commented on test/com/gfredericks/test/chuck/exception_handling_test.cljc:
```diff
@@ -0,0 +1,63 @@
+           (binding [*current-env* (test/empty-env)]
+             (capture-test-var #'this-test-should-fail)
+             [(:report-counters *current-env*) nil]))]
+    #?(:clj (is (not (re-find #"not-falsey-or-exception" out))))
```
does `with-out-str` not work in cljs?

nberger replied:
Yes, it works. It was fixed as part of a01278f

gfredericks commented on src/com/gfredericks/test/chuck/properties.cljc:
```diff
@@ -1,9 +1,10 @@
-            [com.gfredericks.test.chuck.generators :as gen']))
+            [clojure.test.check.properties :as prop
+             #?@(:cljs [:include-macros true])]
+            [com.gfredericks.test.chuck.generators :as gen
```
I'd like to keep the alias here as `gen'`; I use `gen` for `clojure.test.check.generators`.

nberger replied:
Done

gfredericks commented on src/com/gfredericks/test/chuck/clojure_test.cljc:
```diff
@@ -0,0 +1,104 @@
+
+(defn -report
+  [reports]
+  (#?(:clj clojure.test/report :cljs cljs.test/report) reports))
```
If we used a namespace alias for `clojure.test` and `cljs.test` (i.e., the same alias), we wouldn't need the reader conditional here would we?

nberger replied:
Yes. It's done in 39c3faf

gfredericks commented on src/com/gfredericks/test/chuck/clojure_test.cljc:
```diff
@@ -0,0 +1,104 @@
+;;
+;; I think there might be plans for test.check to abstract this logic
+;; into a protocol or something, so I'm not too bothered by the
+;; copypasta for now.
```
did this comment get removed on master? I think it only applied to `not-falsey-or-exception?` which doesn't seem to be what we're using anymore

gfredericks replied:
doesn't look like you addressed this one

nberger replied:
Ups. Taking a look now...

nberger replied:
You are right, it was removed on master. It slipped through here... I just removed it in 8f978c6

nberger commented:
Everything seems to be ok now. I'm going to update the rebased branch with the latest changes and push it here in a few minutes.

nberger commented:
Just pushed the rebased commits.

I kept 7 commits, not just 1, as I think it might be of value to keep those separate in the history. But let me know if you think we should squash some or all of them.

gfredericks commented on src/com/gfredericks/test/chuck/clojure_test.cljc:
```diff
@@ -0,0 +1,98 @@
+
+(defn -testing
+  [name func]
+  (#?(:clj clojure.test/testing :cljs cljs.test/testing) name (func)))
```
whoops I meant to have this changed too, to `ct/testing`

nberger replied:
> whoops I meant to have this changed too, to ct/testing
>
I guessed so, and tried to change it too, but it didn't work. It was
because `testing` is a macro, and this triggers the original problem we had
with trying to refer to macros in different namespaces in clj and cljs...

So, I'm not saying this is impossible to do, but it's not trivial, and it
can be done after merging this PR :D

gfredericks replied:
is it not just because `testing` isn't mentioned in `:refer-macros` at the top? (or that you didn't use `:include-macros true`?)

nberger replied:
> is it not just because testing isn't mentioned in :refer-macros at the
top? (or that you didn't use :include-macros true?

No. AFAIK when there exists a clj ns with the same name as the cljs ns, all
of its macros are available without the need of :refer-macros or anything.
It's explained in the "differences from clojure" page in the cljs wiki, but
I'm with bad connection now and can't find it

nberger replied:
Sorry, we are talking about another ns here... In that case, I tried adding
:refer-macros and it didn't work :)

nberger replied:
> Sorry, we are talking about another ns here... In that case, I tried
adding :refer-macros and it didn't work :)

Or perhaps I didn't :(.

I'll try again tomorrow

nberger replied:
Well, of course you were right, and I was wrong. It worked with `:refer-macros [testing]`. It's done in 23f34e3.

I just checked and what I tried the other day was `[cljs.test :as ct :refer-macros [is testing]]` but then tried to call the macro as `(ct/testing ...)` instead of `(testing ...)`.

nberger commented:
@gfredericks I just updated test.check dependency to 0.8.2

Let me know if you find anything else :)
